library(stringr)

setwd('/Volumes/N1/RNAseq/')

ProcessSamples <- function(seqs, workdir, samples)
{
  for (sample in samples)
  {
    tmp <- read.table(paste(workdir,'/01_output/',sample,'_aligned.FC',sep = ''), header=T, sep='\t',fill=T,stringsAsFactors = F, check.names = F)
    tmp<-tmp[c(1,6,7)]
    names(tmp)<-c('gene','length','reads')
    tmp$RPK <- tmp$reads / (tmp$length /1000)
    PM <- sum(tmp$RPK)  / 1000000
    tmp$TPM <- tmp$RPK / PM
    tmp<-tmp[c(1,5)]
    names(tmp)[2]<-paste(sample,'TPM',sep = '_')
    if (nrow(seqs) == 0)
      seqs <- tmp
    else
      seqs <- merge(seqs,tmp,by='gene', all = T)
  }
  return(seqs)
}

rnaseq <- data.frame(stringsAsFactors = F)
rnaseq <- ProcessSamples(rnaseq, '/Volumes/N1/RNAseq/90-1106610255', c(201,202,203,204,205,206,207,208,210,219,220,221,222,223,225,226,227,228,229,230,231,232,233,234,235,237,238,239,240,241,242,243,244,245,246,247,248,249,250,251,252,253,254,255,256,257,258,259,260,261,262,263,264,265,266,267,268,269,270,271,272,273,274,275,276,277,279,280,282,283,284,285,286,287,288,289,290,291,292,293,294,296,297,298,299,300,301,302,303,304,306,307,308,309,310,311,312,313,314,315,316,317,318,319,320,321,322,323,324,325,326,327,328,329,330,331,332,333,334,335,336,337,338,339,340,341,342,343,344,345,346,347,348,349,350,351,352,353,354,355,356,357,358,359,360,361,362,363,364,365,366,367,368,369,370,371,372,373,374,376,377,378,380,381,382,383,384,385,386,387,388,389,390,391,392,393,394,395,396,397,399,400))
rnaseq <- ProcessSamples(rnaseq, '/Volumes/BURAN/RNAseq/90-1120364334', c(401,402,403,404,405,406,407,408,409,411,412,413,414,415,416,417,418,419,420,421,422,423,424,425,426,427,428,429,430,431,434,435,437,438,439,440,441,442,443,444,445,446,447,448,449,450,451,452,453,454,455,456,457,458,459,460,461,462,463,464,465,466,467,468,469,470,471,472,473,475,476,477,478,480,481,482,483,485,489,490,491,492,493,494,495,496,497,498,499,500,501,502,503,504,505,506,507,508,509,510,511,512,513,514,515,516,517,518,519,520,521,522,523,524,525,526,527,528,529,530,531,532,533,534,535,536,537,538,539,540,541,542,543,544,545,546,547,548,549,550,551,552,554,555,556,557,558,559,560,561,562,563,564,565,566,567,568,569,570,571,572,573,574,576,577,578,580,581,582,583,585,586,587,588,589,590,591,592,593,594,595,596,597,598,599,600))
write.table(rnaseq,file='./RNAseq.reads.tsv', sep='\t', quote = F, row.names = F, col.names = T)

#------------------------------------ END --------------------------------------


